be/src/exec/workgroup/work_group.cpp

```diff
@@ -2,73 +2,241 @@
+            _workgroup_expired_versions.push_back(old_unique_id);
+        }
+    }
+    // install new version
+    _workgroup_versions[wg->id()] = wg->version();
+    _wg_io_queue.add(wg);
 }
 
-WorkGroupPtr WorkGroupManager::pick_next_wg_for_io() {
-    return _wg_io_queue.pick_next();
+void WorkGroupManager::alter_workgroup_unlocked(const WorkGroupPtr& wg) {
+    create_workgroup_unlocked(wg);
 }
 
-WorkGroupQueue& WorkGroupManager::get_cpu_queue() {
-    return _wg_cpu_queue;
+void WorkGroupManager::delete_workgroup_unlocked(const WorkGroupPtr& wg) {
+    auto unique_id = wg->unique_id();
+    auto wg_it = _workgroups.find(unique_id);
+    if (wg_it != _workgroups.end()) {
+        wg_it->second->mark_del();
+    }
 }
 
-WorkGroupQueue& WorkGroupManager::get_io_queue() {
-    return _wg_io_queue;
+std::vector<TWorkGroup> WorkGroupManager::list_workgroups() {
+    std::shared_lock read_lock(_mutex);
+    std::vector<TWorkGroup> alive_workgroups;
+    for (auto it = _workgroups.begin(); it != _workgroups.end(); ++it) {
+        const auto& wg = it->second;
+        if (!wg->is_marked_del() && wg->version() != WorkGroup::DEFAULT_VERSION) {
+            alive_workgroups.push_back(wg->to_thrift());
+        }
+    }
+    return alive_workgroups;
 }
 
-namespace {
-class DefaultWorkGroupInitialization {
-public:
-    DefaultWorkGroupInitialization() {
-        auto default_wg = std::make_shared<WorkGroup>("default_wg", 1, 10, 10, 10, WorkGroupType::WG_DEFAULT);
-        default_wg->init();
-        WorkGroupManager::instance()->add_workgroup(default_wg);
+std::vector<TWorkGroup> WorkGroupManager::list_all_workgroups() {
+    std::vector<TWorkGroup> workgroups;
+    {
+        std::shared_lock read_lock(_mutex);
+        workgroups.reserve(_workgroups.size());
+        for (auto it = _workgroups.begin(); it != _workgroups.end(); ++it) {
+            const auto& wg = it->second;
+            auto twg = wg->to_thrift_verbose();
+            workgroups.push_back(twg);
+        }
     }
-} _;
-} // namespace
+    std::sort(workgroups.begin(), workgroups.end(),
+              [](const auto& lhs, const auto& rhs) { return lhs.name < rhs.name; });
+    return workgroups;
+}
+
+bool IoWorkGroupQueue::try_offer_io_task(WorkGroupPtr wg, const PriorityThreadPool::Task& task) {
+    // Not implemented
+    return true;
+}
+
+void IoWorkGroupQueue::close() {
+    // TODO: Not Implemented
+}
+
+void WorkGroupManager::close() {
+    _wg_io_queue.close();
+}
+
+StatusOr<PriorityThreadPool::Task> WorkGroupManager::pick_next_task_for_io() {
+    // TODO: Not Implemented
+    return Status::OK();
+}
+
+bool WorkGroupManager::try_offer_io_task(WorkGroupPtr wg, const PriorityThreadPool::Task& task) {
+    // TODO: Not Implemented
+    return true;
+}
+
+DefaultWorkGroupInitialization::DefaultWorkGroupInitialization() {
+    auto default_wg = std::make_shared<WorkGroup>("default_wg", WorkGroup::DEFAULT_WG_ID, WorkGroup::DEFAULT_VERSION, 1,
```

Comment:
maybe need call init?

Reply:
add_workgroup calls WorkGroupManager::create_workgroup_unlocked in which wg->init() is invoked.

Reply:
OK